File: web/src/api/claimFlow.ts
import type { InvitePreviewResult } from './types';

/**
 * Pure decision logic for the invite-claim funnel, ported from the Kotlin
 * client (src/commonMain/kotlin/com/kinfolk/portal/screens/claim/ClaimFlow.kt)
 * so vitest can hit every branch.
 *
 * The funnel a brand-new kinfolk walks:
 *   email link -> preview invite -> create account (set password) -> acceptInvite
 * Returning kinfolk (account exists) flip to sign-in mode with the same email.
 * An already-signed-in session auto-accepts only when its email matches the
 * invite; otherwise we say whose invite this is and offer sign-out.
 */

export type ClaimStep =
  | { kind: 'loading' }
  | { kind: 'createAccount'; invitedEmail: string; tribeName: string }
  | { kind: 'autoAccept'; invitedEmail: string }
  | { kind: 'wrongAccount'; invitedEmail: string; currentEmail: string }
  | { kind: 'inviteInvalid'; message: string };

export function stepForPreview(
  preview: InvitePreviewResult,
  signedIn: boolean,
  signedInEmail: string | null,
): ClaimStep {
  switch (preview.status) {
    case 'valid':
      if (!signedIn) {
        return { kind: 'createAccount', invitedEmail: preview.invitedEmail, tribeName: preview.tribeName };
      }
      if (signedInEmail !== null && signedInEmail.toLowerCase() === preview.invitedEmail.toLowerCase()) {
        return { kind: 'autoAccept', invitedEmail: preview.invitedEmail };
      }
      return { kind: 'wrongAccount', invitedEmail: preview.invitedEmail, currentEmail: signedInEmail ?? '' };
    case 'claimed':
      return { kind: 'inviteInvalid', message: 'This invite was already used. Sign in with your email and password instead.' };
    case 'expired':
      return { kind: 'inviteInvalid', message: 'This invite has expired. Ask Auntie to send a fresh one.' };
    case 'revoked':
      return { kind: 'inviteInvalid', message: 'This invite is no longer active. Ask Auntie to send a fresh one.' };
    default:
      return { kind: 'inviteInvalid', message: "We couldn't find this invite. Check the link or ask Auntie to resend it." };
  }
}

/** Returns a user-facing problem, or null when the password is acceptable. */
export function validateNewPassword(password: string, confirm: string): string | null {
  if (password.length < 8) return 'Password needs at least 8 characters.';
  if (password !== confirm) return "Passwords don't match.";
  return null;
}

/**
 * Parses the invite id from a launch URL, matching the Kotlin deep-link
 * parser (src/jsMain DeepLink.js.kt): URL fragment `#/claim/<id>`, pathname
 * `/claim/<id>`, or the `?invite=<id>` query param. The query form is what
 * every invite email sends (`${CLAIM_LINK_BASE_URL}?invite=${id}`), so it
 * must parse here or the claim screen never mounts for invited kinfolk.
 */
export function parseInviteId(url: {
  hash: string;
  pathname: string;
  search: string;
}): string | null {
  const fromHash = parseSegment(url.hash.replace(/^#/, ''), 'claim');
  if (fromHash) return fromHash;
  const fromPath = parseSegment(url.pathname, 'claim');
  if (fromPath) return fromPath;
  const invite = new URLSearchParams(url.search).get('invite');
  return invite && invite.trim().length > 0 ? invite : null;
}

function parseSegment(s: string, prefix: string): string | null {
  const parts = s.replace(/^\/+|\/+$/g, '').split('/');
  if (parts.length >= 2 && parts[0] === prefix && parts[1] && parts[1].length > 0) {
    return parts[1];
  }
  return null;
}
/**
 * O-35: hard client-side bound for the auto-accept call. The callable SDK
 * acquires its context (auth/App Check/messaging tokens) BEFORE its own
 * timeout starts, so a stall there hangs past the 20s CallableTimeoutError
 * and the retry card never shows. Racing the whole call keeps the claim
 * screen honest no matter where the stall lives.
 */
export function withTimeout<T>(promise: Promise<T>, ms: number, label: string): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error(`${label} timed out. Please try again.`)), ms);
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (err: unknown) => {
        clearTimeout(timer);
        reject(err instanceof Error ? err : new Error(String(err)));
      },
    );
  });
}

File: mytribe/src/commonMain/kotlin/com/kinfolk/portal/screens/claim/ClaimFlow.kt
package com.kinfolk.portal.screens.claim

/**
 * Pure decision logic for the invite-claim funnel, kept out of the Composable
 * so commonTest can hit every branch (same pattern as LaunchRouter /
 * StartRoute resolvers).
 *
 * The funnel a brand-new kinfolk walks:
 *   email link → preview invite → create account (set password) → acceptInvite
 * Returning kinfolk (account exists) flip to sign-in mode with the same email.
 * An already-signed-in session auto-accepts only when its email matches the
 * invite; otherwise we tell them whose invite this is and offer sign-out.
 */

/** Parsed result of the public getInvitePreview callable. */
data class InvitePreview(
    val status: String,
    val invitedEmail: String = "",
    val tribeName: String = "",
)

sealed interface ClaimStep {
    /** Preview fetch or acceptInvite in flight. */
    data object Loading : ClaimStep

    /** Signed out + claimable invite: collect a password (or sign in). */
    data class CreateAccount(val invitedEmail: String, val tribeName: String) : ClaimStep

    /** Signed in with the right email: call acceptInvite now. */
    data class AutoAccept(val invitedEmail: String) : ClaimStep

    /** Signed in as someone else: must sign out first. */
    data class WrongAccount(val invitedEmail: String, val currentEmail: String) : ClaimStep

    /** Invite can never be claimed (terminal). */
    data class InviteInvalid(val message: String) : ClaimStep
}

fun stepForPreview(
    preview: InvitePreview,
    signedIn: Boolean,
    signedInEmail: String?,
): ClaimStep = when (preview.status) {
    "valid" -> when {
        !signedIn -> ClaimStep.CreateAccount(preview.invitedEmail, preview.tribeName)
        signedInEmail != null &&
            signedInEmail.lowercase() == preview.invitedEmail.lowercase() ->
            ClaimStep.AutoAccept(preview.invitedEmail)
        else -> ClaimStep.WrongAccount(preview.invitedEmail, signedInEmail ?: "")
    }
    "claimed" -> ClaimStep.InviteInvalid(
        "This invite was already used. Sign in with your email and password instead.",
    )
    "expired" -> ClaimStep.InviteInvalid(
        "This invite has expired. Ask Auntie to send a fresh one.",
    )
    "revoked" -> ClaimStep.InviteInvalid(
        "This invite is no longer active. Ask Auntie to send a fresh one.",
    )
    else -> ClaimStep.InviteInvalid(
        "We couldn't find this invite. Check the link or ask Auntie to resend it.",
    )
}

/**
 * Firebase signals an existing account as EMAIL_EXISTS (REST) or
 * auth/email-already-in-use (JS SDK). Either way: switch to sign-in mode.
 */
fun isEmailAlreadyInUse(message: String?): Boolean {
    val m = message?.lowercase() ?: return false
    return "email_exists" in m || "email-already-in-use" in m || "email already in use" in m
}

/**
 * RULING: "secondary needs email verification as well."
 *
 * `acceptInvite` refuses an invitee whose address is not verified, and mails them
 * a verification link on the way out. That refusal is a step to take, not a
 * failure to retry, so the claim screen has to tell it apart from the OTHER
 * failed-precondition that callable throws for a dead invite.
 *
 * Mirrors `isEmailUnverified` in mytribe/web/src/lib/authErrors.ts.
 */
fun isEmailUnverified(message: String?): Boolean {
    val m = message?.lowercase() ?: return false
    return "failed-precondition" in m && "verif" in m
}
/** Returns a user-facing problem, or null when the password is acceptable. */
fun validateNewPassword(password: String, confirm: String): String? = when {
    password.length < 8 -> "Password needs at least 8 characters."
    password != confirm -> "Passwords don't match."
    else -> null
}

/**
 * #931: what the claim card shows when `getInvitePreview` or `claimInviteSignup`
 * refuses for its rate limit. Same words as portal web's rate-limit copy
 * (`mapAuthError`'s `rateLimit` kind in `mytribe/web/src/lib/authErrors.ts`).
 */
const val CLAIM_RATE_LIMITED_MESSAGE = "Too many tries for now. Wait a few minutes, then try again."

/**
 * The claim card's text for a failed preview or signup call: the rate-limit copy
 * when the callable refused for its limit (the native Android SDK carries the
 * server's "Too many attempts. Try again later." message, the desktop REST
 * client the RESOURCE_EXHAUSTED response body), otherwise the error's own
 * message, otherwise [fallback].
 *
 * Neither refusal is a connection problem, so it must not be told as one:
 * trying again immediately is exactly what will not work.
 */
fun claimErrorMessage(message: String?, fallback: String): String =
    if (com.kinfolk.portal.screens.tribe.isRateLimited(message)) CLAIM_RATE_LIMITED_MESSAGE else message ?: fallback
